components/playlist/playlist_service.cc

```diff
@@ -0,0 +1,431 @@
+  const base::Value* playlist_info =
+      prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  if (!playlist_info) {
+    LOG(ERROR) << __func__ << ": Invalid playlist id for recover: " << id;
+    return;
+  }
+
+  VLOG(2) << __func__;
+  media_file_download_manager_->GenerateMediaFileForPlaylistItem(
+      *playlist_info);
+}
+
+base::FilePath PlaylistService::GetPlaylistItemDirPath(
+    const std::string& id) const {
+  return base_dir_.AppendASCII(id);
+}
+
+void PlaylistService::UpdatePlaylistValue(const std::string& id,
+                                          base::Value value) {
+  prefs::ScopedDictionaryPrefUpdate update(prefs_, kPlaylistItems);
+  auto playlist_items = update.Get();
+  playlist_items->Set(id, base::Value::ToUniquePtrValue(std::move(value)));
+}
+
+void PlaylistService::RemovePlaylist(const std::string& id) {
+  prefs::ScopedDictionaryPrefUpdate update(prefs_, kPlaylistItems);
+  auto playlist_items = update.Get();
+  playlist_items->Remove(id, nullptr);
+}
+
+void PlaylistService::CreatePlaylistItem(const CreatePlaylistParams& params) {
+  VLOG(2) << __func__;
+  const PlaylistInfo info = CreatePlaylistInfo(params);
+  UpdatePlaylistValue(info.id, GetValueFromPlaylistInfo(info));
+
+  NotifyPlaylistChanged(
+      {PlaylistChangeParams::ChangeType::kChangeTypeAdded, info.id});
+
+  base::PostTaskAndReplyWithResult(
+      task_runner(), FROM_HERE,
+      base::BindOnce(&base::CreateDirectory, GetPlaylistItemDirPath(info.id)),
+      base::BindOnce(&PlaylistService::OnPlaylistItemDirCreated,
+                     weak_factory_.GetWeakPtr(), info.id));
+}
+
+void PlaylistService::OnPlaylistItemDirCreated(const std::string& id,
+                                               bool directory_ready) {
+  VLOG(2) << __func__;
+  if (!directory_ready) {
+    NotifyPlaylistChanged(
+        {PlaylistChangeParams::ChangeType::kChangeTypeAborted, id});
+    return;
+  }
+
+  DownloadThumbnail(id);
+  GenerateMediafileForPlaylistItem(id);
+}
+
+void PlaylistService::DownloadThumbnail(const std::string& id) {
+  const base::Value* item_value = prefs_->Get(kPlaylistItems)->FindDictKey(id);
+  DCHECK(item_value);
+  const base::Value* create_params_value =
+      item_value->FindDictKey(kPlaylistCreateParamsKey);
+  DCHECK(create_params_value);
+  const std::string* thumbnail_url =
+      create_params_value->FindStringKey(kPlaylistPlaylistThumbnailUrlKey);
+  if (!thumbnail_url || thumbnail_url->empty()) {
+    VLOG(2) << __func__ << ": thumbnail url is not available.";
+    return;
+  }
+
+  thumbnail_downloader_->DownloadThumbnail(
+      id, GURL(*thumbnail_url),
+      GetPlaylistItemDirPath(id).Append(kThumbnailFileName));
+}
+
+void PlaylistService::OnThumbnailDownloaded(const std::string& id,
+                                            const base::FilePath& path) {
+  DCHECK(IsValidPlaylistItem(id));
+
+  if (path.empty()) {
+    VLOG(2) << __func__ << ": thumbnail fetching failed for " << id;
+    NotifyPlaylistChanged(
```

Comment:
it doesn't look like we are doing anything with this event so if the thumbnail fails to download does it ever retry?